src/modules/python/pbs_hooks/PBS_power.PY

```diff
@@ -320,40 +344,73 @@ if this_event.type == pbs.PERIODIC:
+                        nodes[host]["can_power_off"] = 0
+                    if resv.reserve_state == pbs.RESV_STATE_CONFIRMED:
+                        reserve_start = resv.reserve_start
+                        if vn not in exec_vnodes:
+                            exec_vnodes[vn] = {}
+                            exec_vnodes[vn]["neededby"] = reserve_start
+                        elif reserve_start < exec_vnodes[vn]["neededby"]:
+                            exec_vnodes[vn]["neededby"] = reserve_start
+                        # Do not power off a node if it has reservation starting
+                        # in (time_now + min_node_down_delay + 900) seconds
+                        if time_now > (reserve_start - (min_node_down_delay + 900)):
                             nodes[host]["poweroff"] = 0
+                            nodes[host]["can_power_off"] = 0
 
                 # Is the node idle enough to put to sleep or power down?
-                if vnode.last_used_time:
-                    last_used_time = vnode.last_used_time
-                else:
-                    last_used_time = 0
-                idle_time = time_now - last_used_time
-                if node_idle_limit < idle_time:
-                    nodes[host]["poweroff"] = 1
+                if nodes[host]["can_power_off"]:
+                    if vnode.last_used_time:
+                        last_used_time = vnode.last_used_time
+                    else:
+                        last_used_time = 0
+                    idle_time = time_now - last_used_time
+                    if node_idle_limit < idle_time:
+                        nodes[host]["poweroff"] = 1
 
         # POWER-ON/RAMP-UP check: See if the node is down and
         # needs to be brought up. Check if the node needs to be
         # up before this periodic event runs again.
         if vnode.state == pbs.ND_SLEEP:
+            if power_on_off_enable:
+                # Ignore the nodes in sleep_node_list as they are
+                # being powered up by previous interation of the hook.
+                if vn in sleep_node_list:
+                    continue
+            # Look for upcoming reservation starts in the node.
+            for resid in rs_list:
+                resv = resvlist[resid.lstrip()]
+                if resv.reserve_state == pbs.RESV_STATE_CONFIRMED:
+                    reserve_start = resv.reserve_start
+                    if vn not in exec_vnodes:
+                        exec_vnodes[vn] = {}
+                        exec_vnodes[vn]["neededby"] = reserve_start
+                    elif reserve_start < exec_vnodes[vn]["neededby"]:
+                        exec_vnodes[vn]["neededby"] = reserve_start
             if vn in exec_vnodes and exec_vnodes[vn]["neededby"]:
                 neededby = exec_vnodes[vn]["neededby"]
             else:
                 neededby = 0
-            if neededby and neededby < (time_now + freq):
+            if neededby and neededby < (time_now + 900):
```

Comment:
Better to add the significance of 900 here for better readability.

Reply:
Okay.